DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces_overview/Readme.md

```diff
@@ -0,0 +1,119 @@
+
+![](assets/data_streaming.svg)
+
+### Memory-Mapped Agent (Control/Status Register)
+You may wish to register your data in a memory-mapped register space.
+
+![](assets/data_mm_agent.svg)
+
+There are two ways you can do this:
+- Use (non-pointer) members of your functor. This is the default way to get data into your kernel as demonstrated in the [Naive Design](naive).
+- Use host pipes configured to use the `sycl::ext::intel::experimental::protocol_avalon_mm` protocol.
+
+oneAPI kernels do not have return types, so you must use one of the three data interfaces to pass output data. If you would like to store your output in the CSR for a memory-mapped host device to access (as shown in the graphic above), please review the [csr-pipes](csr-pipes/) example design.
+
+## Sample Structure
+There are 5 different example designs in this sample, all of which implement a simple vector addition. You can compare the C++ source code for each of these designs.
+1. [Naive](naive/) This simple vector add implementation is similar to the one presented in the [fpga_compile](https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/C%2B%2BSYCL_FPGA/Tutorials/GettingStarted/fpga_compile) code sample. It demonstrates a kernel that uses a register-mapped invocation interface, and a single shared memory-mapped host data interface. This is the only design that can support the full-system compilation flow, since the other designs implement interface customizations that are not supported by any BSP.
+2. [Streaming invocation](handshake-control/) This implementation uses a streaming invocation interface and conduits to pass the input pointer arguments instead of the CSR. It uses a single, shared memory-mapped host data interface.
+3. [Customized memory-mapped host data interface](mm-host/) This implementation uses a register-mapped invocation interface, and demonstrates how to use `annotated_arg` to customize the memory-mapped host data interface.
+4. [Streaming Data (host pipes)](host-pipes/) This implementation uses a register-mapped invocation interface, and demonstrates how to use host pipes to configure a streaming data interface.
+5. [CSR Data](csr-pipes/) This implementation uses a register-mapped invocation interface, and demonstrates how to use host pipes to configure a streaming data interface, and how to return output into the CSR using the `protocol_avalon_mm` pipe protocol.
+
+All designs support four compilation options: Emulator, Simulator, Optimization Report, FPGA Hardware. For details on the different compilation options, see [fpga_compile](https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/C%2B%2BSYCL_FPGA/Tutorials/GettingStarted/fpga_compile) tutorial.
+
+For detailed instructions on how build and run each design, see the documentation under each design's directory.
```

Comment:
Because these all share the same build instructions, I think that I'd prefer you having a top level cmake file that selects which sources to use, depending on the options given to cmake.
e.g. cmake .. -DTYPE=[NAIVE/CSR/HANDSHAKE/HOSTPIPES/MMHOST] 

Your top level cmake would have something like this:
```
if(${TYPE} MATCHES NAIVE)
  add_src(naive)
else if(${TYPE} MATCHES CSR)
  add_src(csr-pipes)
...
```

This way, all the build instructions are in this README.


Similarly, I think that I'd prefer all the sublevel README instructions be in this README. I'd rather have one place to compare everything rather than having to go a change folders back and forth to realize which one fits better my application.

Reply:
@yuguen-intel  Why do you want a single top-level cmake file? I think multiple CMake files makes for better consistency, since each sub-sample can act as its own independent mini-project.

I similarly think that putting all the descriptions for each of the sub-projects together in the top-level readme would make it too long. Split up into pieces, it's easier to digest.

Reply:
It avoids the need for doing:
- `cd project 1` 
- `mkdir build`
- `cd build`
- `cmake .. `
- ...
- `cd ../../project 2`
- `mkdir build`
- `cd build`
- `cmake ..`
- ...

one can simply do:
- `mkdir build`
- `cd build`
- `cmake .. -DTYPE=PROJECT1`
- ...
- `cmake .. -DTYPE=PROJECT2`
- ...

This is how other samples that have multiple subprojects are configured

+ all of these cmake files are exactly identical, there could be a single one which just selects different source code.

Reply:
More generally, I think we should discuss where to put all of these new interface samples
It sounds like this is an overview, but William made a detailed one, and we may do more in the future.

So we should think ahead a plan for a nice structure.

We may want a new ipa_interfaces folder, which could contain this nice README, and then multiple individual samples such as the one William wrote, the host pipes one, etc.

Reply:
@yuguen-intel and I discussed these last week; I am also convinced that this is the way (i.e.
1. an ipa_interfaces directory under `features` that contains this and other IPA tutorials
2. modify the CMake file to select the different c++ file depending on a compile flag

Reply:
for number 1. should I be re-grouping the code samples in my PR?

Reply:
no, as we discussed just move yours and the others will follow